bash-preexec.sh

```diff
@@ -47,6 +47,11 @@ if [[ -z "${BASH_VERSINFO-}" ]] || (( BASH_VERSINFO[0] < 3 || (BASH_VERSINFO[0]
     return 1
 fi
 
+# We do not enable bash-preexc in non-interactive shells (except in bats tests)
+if [[ $- != *i* && ${BATS_TEST_FILENAME-} != *.bats ]]; then
```

Comment:
Why not make `bats` run with `-i` instead?

Reply:
What do you mean? Bats doesn't have such an option. I also checked the upstream `bats-core/bats-core` for the latest version, but it doesn't seem to support the `-i` option either. If you mean `bash -i "$(type -P bats)" test`, it doesn't work because `bats` anyway internally spawns `bash` without `-i`.

Reply:
> If you mean `bash -i "$(type -P bats)" test`, it doesn't work because `bats` anyway internally spawns `bash` without `-i`.

Yeah I was thinking about that. Was also looking if there is an equivalent `set` but that also seems to not be there.

Reply:
I've skimmed through my `/etc/profile.d` and the only one different I could find is `colorls.sh`
```shell
[ ! -t 0 ] && return
```

Reply:
> I've skimmed through my `/etc/profile.d` and the only one different I could find is `colorls.sh`
> 
> ```shell
> [ ! -t 0 ] && return
> ```

I have difficulty in understanding the intent of this comment. Do you request me to do something on this PR before this is merged?

Reply:
Just suggesting some alternatives. `bash -i` seems to not be usable within `bats`, but maybe `[ ! -t 0 ]` can achieve similar check for interactive. The only weirdness is the check against `BATS_TEST_FILENAME`, but if it can't be worked around, than it should be fine in the current approach

Reply:
> maybe `[ ! -t 0 ]` can achieve similar check for interactive

Maybe? How much are you sure about this? Did you actually try it?

Reply:
I've only tried it as:
```console
$ [ ! -t 0 ] && echo "Non-interactive" || echo "Interactive"
Interactive
```

My bash knowledge doesn't go deep, and I don't know how to setup a non-interactive environment to test that. As far as I can tell, the nuance is that `-t` checks if it is outputting to a terminal or not, and `i` flag is checking for the existence of the flag or not. Probably in Github CI, `-t` will show as "interactive" while the `i` flag will show as "non-interative", but which is the appropriate result in this case :shrug:

Reply:
Re Github CI, both of them seem to output as `non-interactive`, so that's irrelevant:
https://github.com/LecrisUT/test-interactive/actions/runs/11071448788/job/30763327247

Reply:
Thank you for the explanation. So if I understand it correctly, you just suggested `[ ! -t 0 ]` to check the interactive shell? I haven't yet tried, but I don't see how that solves the issue of `BATS_TEST_FILENAME`.

Reply:
Well we can try if the check gives different results with `[ ! -t 0 ]`. Depends on how `bats` is spawning the environment I guess. The explicit check for `BATS_TEST_FILENAME` feels weird, and probably having an override of `BASH_PREXEC_NO_CHECKS` which would be populated by the `bats` test would be safer. Your call on this.

Also as a non-author and external commenter I cannot resolve discussions so feel free to resolve it when you have tested the other options and decided on one.

Reply:
I now tried `[ ! -t 0 ]` instead of `[[ $- != *i* && ${BATS_TEST_FILENAME-} != *.bats ]]`. As I anticipated, it doesn't work when the `bats` test is triggered in an environment where `stdin` is not connected to TTY/PTY.

Reply:
> The explicit check for `BATS_TEST_FILENAME` feels weird,

Yeah, I actually agree with this.

> and probably having an override of `BASH_PREXEC_NO_CHECKS` which would be populated by the `bats` test would be safer.

That seems cleaner, but I'm not sure if it is possible. I'll try.

Reply:
I confirmed that the variable set at the top level in the `.bats` files is visible from each test.

I updated the commit: [`3e1581c..30e09cb`](https://github.com/rcaloras/bash-preexec/compare/3e1581c41b1b5b0191fb75d5b8684ffdb183ae47..30e09cb48024c6ef027588757c33c6b23b3b64f8).

I first defined `BASH_PREEXEC_SOMETHING` at the top level, but I changed it to `__bp_something` because it is still not a part of a public interface and is not intended to be used by users.

Reply:
Awesome, no notes. Go ahead and resolve this thread.

Reply:
Thanks!